### ISSUE TITLE
[Bug Report] deviceClient.GetFileUploadSasUriAsync throws null ref exception 

### ISSUE THREAD

**Issue by Skaanning** [Azure/azure-iot-sdk-csharp#2323]
# Context

- **OS, version, SKU and CPU architecture used:** Windows 11 and raspberry pi os
![image](https://user-images.githubusercontent.com/4522165/159920942-416aa75d-dcc2-4ad9-a2f7-ca598feac9e7.png)

- **Application's .NET Target Framework :** dotnet6
- **Device:** Laptop & raspberry pi 4
- **SDK version used:** 
      Microsoft.Azure.Devices 1.37.0
      Microsoft.Azure.Devices.Client 1.40.0
      Microsoft.Azure.Devices.Provisioning.Client 1.19.1
      Microsoft.Azure.Devices.Provisioning.Transport.Mqtt 1.17.1

## Description of the issue
Using the `ProvisioningDeviceClient` and a X509Certificate I register my device. I then use the registration result, to create a new `DeviceClient` including the certificate via the DeviceAuthenticationWithX509Certificate. I then try to get a `FileUploadSasUri` from the device client. This however fails deep inside a `HttpClientHelper`. As far as I can see this happens because a flag _usingX509ClientCertificate is false.

fails in the highlighted area
![image](https://user-images.githubusercontent.com/4522165/159928038-87e84fe4-2dee-4cda-946a-4c10e08efb86.png) 

Exception
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Azure.Devices.Client.Transport.HttpClientHelper.ExecuteAsync(HttpMethod httpMethod, Uri requestUri, Func`3 modifyRequestMessageAsync, Func`2 isSuccessful, Func`3 processResponseMessageAsync, IDictionary`2 errorMappingOverrides, CancellationToken cancellationToken)
```

## Code sample exhibiting the issue

```c#
 public async Task SetupIotDeviceClient(X509Certificate2 certificate)
    {
        var registrationResult = await RegisterDeviceWithProvisioningService(certificate);
        var auth = new DeviceAuthenticationWithX509Certificate(registrationResult.DeviceId, certificate);
        var clientOptions = new ClientOptions{ SdkAssignsMessageId = SdkAssignsMessageId.WhenUnset };
        
        var deviceClient = DeviceClient.Create(registrationResult.AssignedHub, auth, TransportType.Mqtt, clientOptions);
        //this throws the error
        var errorHere = await deviceClient.GetFileUploadSasUriAsync(new FileUploadSasUriRequest {BlobName = "hello"});
    }

async Task<DeviceRegistrationResult> RegisterDeviceWithProvisioningService(X509Certificate2 certificate, int retry = 0)
    {
        var securityProvider = new SecurityProviderX509Certificate(certificate);
        var mqtt = new ProvisioningTransportHandlerMqtt();
        var provisioningDeviceClient = ProvisioningDeviceClient.Create(GlobalDeviceEndpoint, _config.DeviceProvisioningScope, securityProvider, mqtt);

        Log.Information("Registering device {RegistrationId} and assigning to an iot hub", securityProvider.GetRegistrationID());
        
        var result = await provisioningDeviceClient.RegisterAsync();
        if (result.Status == ProvisioningRegistrationStatusType.Assigned)
            return result;
       
        // handle not provisioned stuff down here
    }
```

**Comment by Skaanning** [comment]
@drwill-ms I dont see how they could be older, I use nuget.org as source and this is my csproj file. 

![image](https://user-images.githubusercontent.com/4522165/161013111-aad7f8c2-0d13-4ded-9597-4b0422266dc3.png)

I tried cleaning and rebuilding everything, i still get the error

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Azure.Devices.Client.Transport.HttpClientHelper.ExecuteAsync(HttpMethod httpMethod, Uri requestUri, Func`3 modifyRequestMessageAsync, Func`2 isSuccessful, Func`3 processResponseMessageAsync, IDictionary`2 errorMappingOverrides, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.HttpClientHelper.PostAsync[T1,T2](Uri requestUri, T1 entity, IDictionary`2 errorMappingOverrides, IDictionary`2 customHeaders, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.HttpTransportHandler.GetFileUploadSasUriAsync(FileUploadSasUriRequest request, CancellationToken cancellationToken)
   at dol.Wireless.CloudAgent.Setup.IoTClientBuilder.UploadFileAsBlob(DeviceClient deviceClient, String blobName, Byte[] file) in C:\projects\dol.Wireless.IoTAgent\src\dol.Wireless.CloudAgent\Setup\IoTClientBuilder.cs:line 153
```

**Comment by drwill-ms** [comment]
> The image I attached before is from jetbrains rider's decompiled idea of what is going on in that HttpClientHelper class

Ah, that explains why it looked different than what I was seeing.

**Comment by drwill-ms** [comment]
I've been able to repro using Windows Subsystem for Linux. I'll need to debug to identify the root cause.

**Comment by Skaanning** [comment]
Hey @drwill-ms. 
Is this something that can be fixed soon or are we talking months and months before its updated? Just so I know if I should make a workaround or not.

**Comment by drwill-ms** [comment]
@Skaanning I've been able to find the root cause. It seems we have some faulty logic for using the cert specified to connect for file upload operations and a matching end-to-end test that didn't consider more conditions. I'm going to get a fix out before too long (meaning probably by next week unless we encounter trouble with releasing), but I do have a work around for you in the meantime, which may help.

It turns out that if you don't specify a `ClientOptions` in the call to create a `DeviceClient`, the certificate is properly shared with the HTTP transport layer.

Thank you for finding and notifying us of this error. I apologize for the inconvenience.

**Comment by drwill-ms** [comment]
Oh, the other work around would be to explicitly set the value in `ClientOptions.FileUploadTransportSettings.ClientCertificate`.

**Comment by Skaanning** [comment]
Thanks for the help and updates

**Comment by drwill-ms** [comment]
The fix for this has shipped with version 1.41.0.

<https://github.com/Azure/azure-iot-sdk-csharp/releases/tag/2022-5-6>
